### DISCUSSION TITLE
compare value before appending to to file

### DISCUSSION THREAD

**Discussion by kaitian521** [Tencent/MMKV#1143] [Ideas]
In my app, there exists a kind of blocks or ANRs.

Some keys are set to MMKV files frequently, but in many cases, the value for the key seldom change. For example, cloud configs from servers, some custom settings  from user.

If a key is inserted to a file many times, it can lead to rewrite of file which can cause IO blocks

I spend some time to writing codes, compare values before set, just like below code

```
// newPersonId change not so frequent in my app
public static void updatePersonId(String newPersonId) {
    String nowPersonId = getPersonId();
    // 抑制MMKV重写
    if (!TextUtils.equals(newPersonId, nowPersonId)) {
        mmkv.encode("persion_id", newPersonId);
    }
}
```

Then blocks here almost disappear in my app after compare values

so, would MMKV like to add some supports for this circumstances?
eg, consider adding a dozen of setX functions like:
`setIntWithCompare` `setStringWithCompare` `setLongWithCompare`

Thanks

===================== my tests ========================
```
auto mmkv = [MMKV mmkvWithID:@"test0"];
NSString *key = [NSString stringWithFormat: @"HelloWorld!"];
NSString *value = [NSString stringWithFormat:
@"There are two roads in the forest \
  One is straight and leads to the light \
  The other is crooked and full of darkness \
  Which one will you choose to walk? \
\
  The straight road may be easy to follow \
  But it may lead you to a narrow path \
  The crooked road may be difficult to navigate \
  But it may open up a world of possibilities \
\
  The choice is yours to make \
  Decide wisely and with a open heart \
  Walk the path that leads you to your dreams \
  And leaves you with no regrets at the end of the day"];
double start = [[NSDate date] timeIntervalSince1970] * 1000;
for(int i = 0; i < 300000; i++) {
    /**
     * 判断值是否相同再写入
     */
    if ([mmkv getValueSizeForKey:key actualSize:true] != [value length] 
      || ![value isEqualToString: [mmkv getStringForKey: key]]) {
        [mmkv setString: value forKey:key];
    }
}
double end = [[NSDate date] timeIntervalSince1970] * 1000;
NSLog(@"funcionalTest, time = %f", (end - start));
```

<img width="604" alt="image" src="https://github.com/Tencent/MMKV/assets/1303341/ca90274f-ac1b-4efc-8d38-ea0a02298d1d">

In addition, the time of running gaps will be exaggerated in production environment: user's phone

**Comment by lingol** [comment]
Interesting idea.

**Comment by lingol** [comment]
Released with [v1.3.2](https://github.com/Tencent/MMKV/releases/tag/v1.3.2).
